[H1] feat: get read count

baksohyeon opened the pull request:
- 문제
     - https://github.com/remzi-arpacidusseau/ostep-projects/tree/master/initial-xv6
- 참고 문서:
     - https://github.com/simnalamburt/snucse/blob/main/Operating%20Systems/note.md
     - https://github.com/baksohyeon/xv6-riscv/blob/riscv/docs/0001.intro-to-xv6-system-calls.md
     - https://clownote.github.io/2021/02/27/xv6/Xv6-operating-system-organization/
## Xv6 organization

Xv6 kernel source files:

| File              | Description                                      |
|-------------------|--------------------------------------------------|
| bio.c             | Disk block cache for the file system.            |
| console.c         | Connect to the user keyboard and screen.         |
| entry.S           | Very first boot instructions.                    |
| exec.c            | exec() system call.                              |
| file.c            | File descriptor support.                         |
| fs.c              | File system.                                     |
| kalloc.c          | Physical page allocator.                         |
| kernelvec.S       | Handle traps from kernel, and timer interrupts.  |
| log.c             | File system logging and crash recovery.          |
| main.c            | Control initialization of other modules during boot. |
| pipe.c            | Pipes.                                           |
| plic.c            | RISC-V interrupt controller.                     |
| printf.c          | Formatted output to the console.                 |
| proc.c            | Processes and scheduling.                        |
| sleeplock.c       | Locks that yield the CPU.                        |
| spinlock.c        | Locks that don’t yield the CPU.                  |
| start.c           | Early machine-mode boot code.                    |
| string.c          | C string and byte-array library.                 |
| swtch.S           | Thread switching.                                |
| syscall.c         | Dispatch system calls to handling function.      |
| sysfile.c         | File-related system calls.                       |
| sysproc.c         | Process-related system calls.                    |
| trampoline.S      | Assembly code to switch between user and kernel. |
| trap.c            | Code to handle and return from traps and interrupts. |
| uart.c            | Serial-port console device driver.               |
| virtio_disk.c     | Disk device driver.                              |
| vm.c              | Manage page tables and address spaces.           |


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a new system call, getreadcount(), allowing user programs to retrieve the total number of read operations performed system-wide.
  * Added a user test program to verify the correctness and concurrency safety of the getreadcount() system call.

* **Bug Fixes**
  * None.

* **Documentation**
  * Expanded and clarified the system call documentation, including detailed explanations of xv6 internals, trap handling, and synchronization primitives.
  * Added descriptive comments to user library functions for improved clarity.

* **Style**
  * Added inline comments and improved code readability in several files without altering logic.

* **Chores**
  * Updated build scripts to include the new test program in the build process.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

baksohyeon committed:
comment: description

baksohyeon committed:
feat: get read count

simnalamburt commented on kernel/sysfile.c:
```diff
@@ -16,6 +16,8 @@
 #include "file.h"
 #include "fcntl.h"
 
+int readcount = 0;
```
여러 코어가 동시에 `readcount` 에 접근할경우 data race가 발생합니다.

baksohyeon replied:
앗 그렇네요! 여러 코어에서 동시에 접근하면 진짜 data race 날 수 있겠어요.
그 부분은 제가 놓쳤던 것 같아요… 알려주셔서 감사합니다!
thread-safe하게 바꿔볼 수 있도록 다시 한번 고민해보겠습니다 🙇‍♀️

baksohyeon replied:
아토믹으로 수정 완료했습니다!

simnalamburt reviewed (changes requested):
`kernel/types.h` 에 `typedef unsigned int uint;` 라는 같은 문장이 두번 쓰여져있어요

baksohyeon committed:
feat: prevents data loss from race conditions when multiple cores access readcount simultaneously

  1. Added spinlock: struct spinlock readcountlock at kernel/sysfile.c:20
  2. Protected increment: sys_read() now uses acquire(&readcountlock) before incrementing and release(&readcountlock) after
  3. Protected read: sys_getreadcount() now safely reads the value under lock protection
  4. Initialization: Added readcountinit() function and called it from main.c:51 during kernel boot

baksohyeon committed:
test: adds multicore stress test

Introduces a new test case to evaluate the system's ability to handle concurrent read operations from multiple processes.

This test simulates a multicore environment by creating several worker processes, each reading from a file concurrently.
The test verifies that the total number of reads performed across all processes is accurately tracked, ensuring the system correctly handles contention and synchronization issues.

simnalamburt commented on kernel/sysfile.c:
```diff
@@ -78,14 +85,20 @@ sys_read(void)
   argint(2, &n);
   if(argfd(0, 0, &f) < 0)
     return -1;
+  acquire(&readcountlock);
```
스핀락으로 바꾸신거 좋긴한데, https://en.cppreference.com/w/c/language/atomic.html 아토믹을 쓰시면 훨씬 쉽고 더 빨라용

baksohyeon replied:
둘 다 개념 정리가 좀 더 필요한 것 같아요.
다시 한 번 꼼꼼히 살펴보고 말씀대로 처리해보겠습니다! 
감사합니다!

baksohyeon replied:
수정 완료했습니닷 감사합니다!

baksohyeon committed:
wip

baksohyeon committed:
fix: remove unused proc property

baksohyeon committed:
docs:: add document

baksohyeon committed:
refactor: read count to atomic integer

Updates the read count variable to an atomic integer.
Removes the spinlock used for protecting the read count,
as atomics provide thread-safe increment and access.
This simplifies the code and improves performance by avoiding
unnecessary locking.

baksohyeon commented:
[코드 리뷰 관련 피드백] 
이번 PR에서 시스템콜 관련 핵심 diff는 17줄 정도였고, 나머지는 주석이나 whitespace 등 불필요한 변경이 많았습니다. 
리뷰 효율을 떨어뜨릴 수 있다는 피드백을 받았고, 앞으로는 실제 변경과 무관한 수정은 별도 PR로 분리하겠습니다. 
이후 작업부터는 의도에 맞게 정돈된 PR을 올리겠습니다.
